# EQUAÇÕES, INEQUAÇÕES E FUNÇÕES#

De acordo com o dicionario matemático disponível em "http://www.somatematica.com.br/dicionarioMatematico" temos que:

**Equação:** Expressão algébrica indicada por uma igualdade, onde há valores desconhecidos expressos por letras (incógnitas). Logo, todo conjunto de expressões no qual há uma igualdade cuja(s) incógita(s) satisfaçam a um conjunto limitado de soluções, então temos uma equação. 

Ex: $x + 2 =0$, $xy - 2x = 2$, $x^2 + y^2 =2^2$.

**Inequação:** Desigualdade verificada a determinado(s) valor(es) atribuídos à variável. Uma inequação é uma equação no qual há uma desigualdade do tipo <, >, >=, =< ou #. 

Ex: $x + 2 > 0,xy - 2x < 2, x^2 + y^2 + 2^2 $.

**Função:** É uma correspondência unívoca entre dois conjuntos em que a cada elemento do primeiro conjunto corresponde a um e somente um elemento do segundo. Dessa forma, temos que uma função é uma relação entre das variáveis, sendo uma dependente e outra independente. 

Ex: $y(x) = x + 2, z = xy - 2x, f(x) = x^2 + y^2$


###EQUAÇÕES###

In [1]:
x = 0

x^2-3*x+2 

2

In [2]:
# Teorema de Pitágoras

c = 2; 
b = 1;
a = sqrt(b^2 + c^2)

2.23606797749979

<span style="font-weight: bold;">Cálculo de uma expressão simbolica<br></span>

In [23]:
using SymPy

In [24]:
x = 4

SymPy.eval(x^2-2*x+1)

9

In [25]:
x = 5

SymPy.eval(x^2-2*x+1)

16

###INEQUAÇÕES###

In [28]:
x,y = Sym("x,y")

(x,y)

In [39]:
ineq1 = x - 5 >= 5

ineq2 = x + y <= 0;

###FUNÇÕES###
**Simbólicas, Genéricas e Anônimas**

**FUNÇÕES MATEMÁTICAS SIMBÓLICAS** 

** Função de uma variável**

In [ ]:
using SymPy

x = Sym("x")

In [65]:
f = x^2 + 1

 2    
x  + 1

In [66]:
display(f)

 2    
x  + 1

In [67]:
typeof(f)

SymPy.Sym

In [68]:
# observe a criação das funções f e F

whos()

                          Base  27310 KB     Module
                       BinDeps    206 KB     Module
                    ColorTypes    223 KB     Module
                        Colors    554 KB     Module
                        Compat    190 KB     Module
                         Conda     64 KB     Module
                          Core   3290 KB     Module
             FixedPointNumbers     33 KB     Module
               FixedSizeArrays    130 KB     Module
                        IJulia    554 KB     Module
                IPythonDisplay     48 KB     Module
                          JSON    201 KB     Module
                 JuliaBoxUtils   8972 bytes  Module
                    MacroTools    162 KB     Module
                          Main  36784 KB     Module
                        Nettle     42 KB     Module
                         Plots   1536 KB     Module
                        PyCall    425 KB     Module
                      Reexport   3628 bytes  Module
            

Observe que o comando Whos() informa que "F" e "f" são diferentes. "F" é uma função e "f" é uma expressão simbólica

In [69]:
display(f(2))

display(f(pi))

display(float(f(pi)))

display(N(f(pi)))

5

      2
1 + pi 

10.869604401089358

10.869604401089358

In [ ]:
g = x^3 + 1

In [ ]:
g + f

** Função de duas variáveis f(x,y)**

In [2]:
y = Sym("y")

LoadError: LoadError: UndefVarError: Sym not defined
while loading In[2], in expression starting on line 1

In [3]:
fxy = x*y - x + y

LoadError: LoadError: UndefVarError: x not defined
while loading In[3], in expression starting on line 1

In [4]:
fxy(0,1)

LoadError: LoadError: UndefVarError: fxy not defined
while loading In[4], in expression starting on line 1

In [ ]:
gxy = x^3 + y^3 + x*y

<strong>Calculado uma função a partir de vetor</strong>

* Usando o comando map

In [80]:
using SymPy

x = Sym("x")

fsimv = sqrt(2*x + cos(x)^3)/sin(x^2+1)

   _______________
  /          3    
\/  2*x + cos (x) 
------------------
      / 2    \    
   sin\x  + 1/    

In [83]:
x = 0:0.01:2*pi

@time map(fsimv,x);

  2.846281 seconds (134.62 k allocations: 5.447 MB)


In [84]:
# utilizando uma função anonima para calcular os valores de 0 a 2*pi somar os resultados

mapreduce(fsimv, +, 0:0.01:2*pi)

  1                      
------ + 686.710479177063
sin(1)                   

* Usando o comando Eval

In [201]:
x = 0:0.01:2*pi

@time SymPy.eval(sqrt(2.*x + cos(x).^3)./sin(x.^2+1));

  0.000208 seconds (24 allocations: 40.516 KB)


* Função direta

In [202]:
x = 0:0.01:2*pi

@time sqrt(2.*x + cos(x).^3)./sin(x.^2+1);

  0.000200 seconds (24 allocations: 40.516 KB)


**FUNÇÕES GENÉRICAS**

As funções genéricas não apresentam variáveis simbolicas definidas pelo SymPy. Usam a seguinte sintaxe:
**nome_funcao**(variável ou variáveis) = **expressão_da_função**

Sendo a "expressão_da_função" uma expressão em função de uma ou mais variáveis.

In [54]:
# criando a função genérica f(x)

f(x) = x^2 + 2*x + 1

f (generic function with 1 method)

In [55]:
# criando a função genérica g(x)

g(x) = x^3 + 2*x^2 + x + 1

g (generic function with 1 method)

In [56]:
# criando a função genérica h(r,s)

h(r,s) = r*s + r + s

h (generic function with 1 method)

In [58]:
# verificando o tipo de dado da funções criadas: tipo função

typeof(f), typeof(g), typeof(h)

(Function,Function,Function)

In [61]:
# verificando o tipo de dado das variáveis criadas
# observe que f,g e h são do tipo function

whos()

                          Base  29013 KB     Module
                          Core   5041 KB     Module
                      LastMain  35549 KB     Module
                          Main   5043 KB     Module
                             f    595 bytes  Function
                             g    607 bytes  Function
                             h    614 bytes  Function


In [62]:
# testando as funções genéricas criadas passando um valor

display(f(2))

display(f(pi))

display(g(2))

display(h(0,1))

display(h(1,pi));

9

17.152789708268944

19

1

7.283185307179586

Obs: Não é possível operar funções genéricas indefinidas, ou seja, sem um valor especificado. Só é possível se x for simbólico definido pelo SymPy. O calculo abaixo ocorrerá erro

In [63]:
f(x) + g(x), f + g

LoadError: LoadError: UndefVarError: x not defined
while loading In[63], in expression starting on line 1

In [65]:
# agora um valor foi repassado, então será possível realizar o calculo

f(2) + g(3)

58

Exemplo de Função de duas variáveis definida

In [ ]:
f(x,y) = 9*x*y - 2*x + y

In [ ]:
f(2,pi)

Forma Equivalente ao for usando a função "map". A função "map" vai cacular o valor de h(x) de 1 a 5.

In [ ]:
h(x) = x^2+1

map(h, 1:5)

Neste outro exemplo, a função "map" vai cacular o valor de h(x) de 1 a 5 e somar os resultados

In [ ]:
h(x) = x^2+1

mapreduce(h, +, 1:5)

** Transformando uma string em uma função genérica**

In [70]:
@eval fsg(x) = $(parse("x^2-2*x-1"))

fsg (generic function with 1 method)

In [72]:
fsg(0)

-1

** Função Genérica do tipo FUNCTION **

In [ ]:
"""
Metodo da Bisseccao

Calcula uma aproximação para uma raiz da função de f(x)

Entre os intervalo [ao,bo] e a tolerencia de erro daddo por tol.

ex bissec("x^2 - 2*x - 8",-5,5,0.000001)
"""

function bissec(funcao,a,b,tol)    

    x = 1
    ao = a
    bo = b
    toll = tol
    interacao = 0
    println("f(x) = ",funcao)
    println()
    
    @eval g(x) = $(parse(funcao)) # transforma a string em uma formula manipulável
    
    if (g(ao)*g(bo)<0)
        
        while (abs(bo-ao) > toll)  &&  (interacao<10000000)
            x=(ao+bo)/2
            interacao = interacao+1
            if g(x)*g(ao) > 0
                ao=x
            else
                bo=x
            end
        end
        println("Interações: ",interacao)
        println("Valor da raiz: ",x)
        println("Valor de f(x)= ",round(g(x),8)," com erro menor que ",toll)
    else
        println("Não há raízes no intervalo, defina outro intervalo")
    end    
end

In [ ]:
@time bissec("sin(x)*x^2+cos(x)*x^2+log(x)*x+cos(sin(x))^2",1,5,0.000000001)

O texto de ajuda entre aspas pode ser acessado pelo comando help (?) 

In [ ]:
? bissec

** Importando uma função no formato de arquivo**

In [ ]:
# É interessante salvar o arquivo da função com o mesmo nome da função

include("fbissec.jl")

In [ ]:
? fbissec

In [ ]:
@time fbissec("sin(x)*x^2+cos(x)*x^2+log(x)*x+cos(sin(x))^2",1,5,0.000000001)

<body><strong>Função Fibonacci</strong><br>
    <p>A Sequência de Fibonacci consiste em uma sucessão de números inteiros, de
      tal forma que, partindo dos dois primeiros números da sequência 0 e 1, os
      números posteriores serão obtidos por meio da soma dos seus dois
      antecessores. Dessa forma, teremos a sequencia: <br>
      <br>
      Termo:0, 1, 2, 3, 4, 5, 6, 7 ,&nbsp; 8 ,&nbsp;&nbsp; 9 ,&nbsp; 10,&nbsp;
      11,&nbsp; 12 ,&nbsp; 13 ,...<br>
      <br>
      Valor_:0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233,...<br>
      <br>
      A função abaixo utiliza a forma recursiva de Fibonacci para encontrar os
      termos da sequencia. A forma recursiva é utilizada em testes de Benchmark
      para testar a velocidade de processamento de uma linguagem de programação,
      de tal forma que quanto menor o tempo para calcular um termo, melhor o
      desempenho da linguagem neste procedimento. <br>
      <br>
      Comparação com outras linguagens utilizando mesmo algoritmo para encontrar
      o 35 termo:</p>
    <table style="width: 674px; height: 54px;" border="1">
      <tbody>
        <tr>
          <td style="width: 81.697px;">Julia:</td>
          <td style="width: 98.067px;">Octave</td>
          <td style="width: 125.067px;">SageMath</td>
          <td style="width: 99.5px;">Scilab</td>
          <td style="width: 175.6px;">Maxima</td>
          <td style="width: 170.7px;">MatLab</td>
        </tr>
        <tr>
          <td><br>
          </td>
          <td><br>
          </td>
          <td><br>
          </td>
          <td><br>
          </td>
          <td><br>
          </td>
          <td><br>
          </td>
        </tr>
      </tbody>
    </table>
  </body>

In [73]:
function fiboR(n) 
    if n < 2 
        return n
    else
        fiboR(n-1) + fiboR(n-2)
    end
end

fiboR (generic function with 1 method)

In [74]:
@time fiboR(35)

9227465

0.116549 seconds (843 allocations: 47.604 KB)


In [ ]:
include("ffiboR_JL.jl")

In [ ]:
? ffiboR_JL

In [ ]:
@time ffiboR_JL(40)

** Forma "Julitronica" de criar IF...Else...End para a função na forma:**  

$~
f(x) = \begin{cases}
\cos(x) & x \geq 0\\
1 - e^{-1/x^2} & \text{x < 0}.
\end{cases}
~$

In [77]:
# função(variável)  = condiçao1 ? comandoA : senao comandoB 
# Significa que: se condição "1" satisfeita, então comandoA, senão comandoB
# no código abaixo, temos que:
# se x>=0 então imprima o calculo de g = cos(x) = ",cos(x), senão imprima o cálculo de g = exp(-1/x^2) 

g(x) = x >= 0  ?  print("g = cos(x) = ",cos(x)) : print("g = exp(-1/x^2) = ",exp(-1/x^2))

g (generic function with 1 method)

In [79]:
# passando um valor de x maior que 0

g(2)

g = cos(x) = -0.4161468365471424

In [81]:
# passando um valor de x menor que 0

g(-1)

g = exp(-1/x^2) = 0.36787944117144233

** Calculando a função a partir de um vetor **

In [209]:
workspace()

In [70]:
x = 0:0.01:2*pi

fgv(x) = sqrt(2*x + cos(x)^3)/sin(x^2+1)

fgv (generic function with 1 method)

* Usando o comando map

In [71]:
@time map(fgv,x);

  0.005193 seconds (2.91 k allocations: 100.006 KB)


In [78]:
# utilizando uma função anonima para calcular os valores de 0 a 2*pi somar os resultados

mapreduce(fgv, +, 0:0.01:2*pi)

687.8988742828413

* Comando eval

In [76]:
x = 0:0.01:2*pi
fgV(x) = sqrt(2.*x + cos(x).^3)./sin(x.^2+1)

@time eval(fgV(x));

  0.004214 seconds (2.96 k allocations: 185.294 KB)


In [73]:
x = 0:0.01:2*pi
fgV(x) = sqrt(2.*x + cos(x).^3)./sin(x.^2+1)

@time fgV(x);

  0.006311 seconds (2.96 k allocations: 185.810 KB)


** FUNÇÕES ANÔNIMAS **

São Funções criadas anonimamente, ou seja, sem um nome para a função. Funções anônimas não precisam usar o SymPy ou a forma de função genérica (funcao(variavel) = expressao). São escritas na forma:

variavel ->  expressao

In [13]:
fa = (x -> x^3-2*x^2-1)

(anonymous function)

In [14]:
fa(0)

-1

<strong>Calculando a função a partir de um vetor</strong>

* Usando o comando map

In [86]:
x = 0:0.01:2*pi

fanx = (x -> sqrt(2*x + cos(x)^3)/sin(x^2+1))

(anonymous function)

In [87]:
@time map(fanx,x);

  0.004017 seconds (6.33 k allocations: 105.780 KB)


In [88]:
# utilizando uma função anonima para calcular os valores de 0 a 2*pi somar os resultados

mapreduce(fanx, +, 0:0.01:2*pi)

687.8988742828413

* Utilizando o Eval

In [16]:
funan = x -> sqrt(2.*x + cos(x).^3)./sin(x.^2+1)
x = 0:0.01:2*pi

@time eval(funan(x));

  0.003547 seconds (58 allocations: 42.769 KB)


In [25]:
funan = x -> x.^3-2.*x.^2-1
x = 0:0.1:1

@time funan(x);

  0.002817 seconds (36 allocations: 2.711 KB)


** QUAL É O MELHOR? FUNÇÃO GENÉRICA, ANÔNIMA OU SIMBÓLICA? **

Vamos fazer um benchmark das funções utilizando 3 métodos: map, eval e direto

**Usando o comando map**

In [48]:
using SymPy
x = Sym("x")

fsimb = -2*x^3 - x^2 + 2*x + 8

     3    2          
- 2*x  - x  + 2*x + 8

In [49]:
fgen(w) = -2*w^3 - w^2 + 2*w + 8

fgen (generic function with 1 method)

In [50]:
fanon = (t ->-2*t^3 - t^2 + 2*t + 8)

(anonymous function)

In [51]:
x = w = t = 0:0.001:10

x , w , t

(0.0:0.001:10.0,0.0:0.001:10.0,0.0:0.001:10.0)

In [55]:
# função simbólica. Retire o ";" para ver os resultados

@time map(fsimb,x);

 37.995546 seconds (2.14 M allocations: 86.603 MB, 0.17% gc time)


In [57]:
# função genérica. Retire o ";" para ver os resultados

@time map(fgen,w);

  0.000703 seconds (20.01 k allocations: 390.984 KB)


In [58]:
# função anonima. Retire o ";" para ver os resultados

@time map(fanon,t);

  0.006040 seconds (70.02 k allocations: 1.145 MB)


Vantagem para a função genérica usando o comando **map**

** Usando o comando eval **

In [42]:
workspace()

In [29]:
using SymPy
x = Sym("x");

In [60]:
# função simbólica. Retire o ";" para ver os resultados

fsimb = -2.*x.^3 - x.^2 + 2.*x + 8

x = 0:0.001:10

@time SymPy.eval(fsimb);

  0.000004 seconds (4 allocations: 160 bytes)


In [62]:
# função genérica. Retire o ";" para ver os resultados

fgen(w) = -2.*w.^3 - w.^2 + 2.*w + 8

w = 0:0.001:10

@time eval(fgen(w));

  0.006428 seconds (1.97 k allocations: 564.841 KB)


In [63]:
# função anonima. Retire o ";" para ver os resultados

fanon = (t ->-2.*t.^3 - t.^2 + 2.*t + 8)

t = 0:0.001:10

@time eval(fanon(t));

  0.004723 seconds (49 allocations: 471.384 KB)


Vantagem para a função simbolica usando o eval

** Usando o comando direto **

In [1]:
workspace()

In [3]:
using SymPy
x = Sym("x");

In [36]:
# função simbólica. Retire o ";" para ver os resultados

x = 0:0.001:10

@time -2.*x.^3 - x.^2 + 2.*x + 8;

  0.001161 seconds (21 allocations: 469.484 KB)


In [48]:
# função genérica. Retire o ";" para ver os resultados

fgen(w) = -2.*w.^3 - w.^2 + 2.*w + 8

w = 0:0.001:10

@time fgen(w);

  0.004027 seconds (1.97 k allocations: 564.653 KB)


In [46]:
# função anonima. Retire o ";" para ver os resultados

t = 0:0.001:10
fan = (t -> -2.*t.^3 - t.^2 + 2.*t + 8)

@time fan(t);

  0.004196 seconds (49 allocations: 471.384 KB)


<span style="font-weight: bold;">Resultados</span><br>
    <br>
<table style="width: 521px;" border="1">
      <tbody>
        <tr align="left">
          <td><br>
          </td>
          <td style="height: 20px; text-align: center;"><span style="font-weight: bold;">map<br>
            </span></td>
          <td style="text-align: center;"><span style="font-weight: bold;">eval<br>
            </span></td>
          <td style="text-align: center;"><span style="font-weight: bold;">direto<br>
            </span></td>
        </tr>
        <tr align="left">
          <td><span style="font-weight: bold;">Simbólica<br>
            </span></td>
          <td style="width: 137.05px;"> 37.995546</td>
          <td style="width: 137.8px;">0.000004</td>
          <td style="width: 139.05px;"> 0.001267</td>
        </tr>
        <tr align="left">
          <td style="width: 100px;"><span style="font-weight: bold;">Genérica<br>
            </span></td>
          <td>0.000703</td>
          <td>0.006428</td>
          <td>0.006702</td>
        </tr>
        <tr>
          <td style="text-align: left;"><span style="font-weight: bold;">Anônima<br>
            </span></td>
          <td style="text-align: left;">0.006040</td>
          <td style="text-align: left;">0.004723</td>
          <td style="text-align: left;">0.004196</td>
        </tr>
      </tbody>
</table>

## CALCULOS DIVERSOS COM EQUAÇÕES, INEQUAÇÕES E FUNÇÕES##

In [ ]:
using SymPy

x,y = Sym("x,y")

** Expansão dos termos de uma expressões **

In [ ]:
y = (x-2)^4 * (x-3)^3

In [ ]:
SymPy.expand(y)

<strong>Fatorização</strong>

In [ ]:
y = x^7 - 17*x^6 + 123*x^5 - 491*x^4 + 1168*x^3 - 1656*x^2 + 1296*x - 432

In [ ]:
SymPy.factor(y)

<strong>Frações parciais</strong>

In [ ]:
fx = (2*x^2 +3*x -1) / (x^2 - 3*x + 2)

In [ ]:
# fraçõs parciais

SymPy.apart(fx)

<strong>Simplificação de expressões</strong>

In [ ]:
frac_fx = 13/(x-2) + -4/(x-1) +2

In [ ]:
SymPy.simplify(frac_fx)

** Substituir variável por um valor ou outra variável**

Para substituir o valor de X na equação acima, utilizamos o comando **subs(variável, variável_expressão, variavel/valor)**

Observe no exemplo 1, que a expressão de Y continua a mesma, só o valor de X que é alterado para o valor (2) e o cálculo é realizado (Y = 2). No exemplo 2, a variável X é alterada para W. E no exemplo 3, as variáveis X e W são substituídas por valores e outras variáveis

* Exemplo 1: substituição por um valor numérico

In [ ]:
using SymPy

x,y = Sym("x,y")

In [ ]:
# x = 0

SymPy.subs(x^2-3*x+2,(x,0))

In [ ]:
# julia não substitui um vetor x = [1, -1, 0]

SymPy.subs(x^2-3*x+2,(x,[1, -1, 0]))

* Exemplo 2: substituição por uma variável

In [ ]:
# x = w

w = Sym("w")

y_w = SymPy.subs(x^2-3*x+2,x,w)

* Exemplo 3: Para substituir mais de uma variável:

In [ ]:
# Substitui x e w por 2 e 1 e efetua o cálculo

r = x*w+2

SymPy.subs(r,(x,2),(w,1)), SymPy.subs(r,x=>2,w=>1)

In [ ]:
# Substitui x e w por p e q e efetua o cálculo

p, q = Sym("p,q")

SymPy.subs(r,(x,p),(w,q)), SymPy.subs(r,x=>p,w=>q)

Resetar o valor das variáveis X e Y